### DISCUSSION TITLE
Setting up Doodad on Compute Canada

### DISCUSSION THREAD

**Discussion by Neo-X** [montrealrobotics/doodad#2] [Q&A]
These are the steps to using Doodad on Compute Canada.

1. Make sure the code works on your local computer
  a) Download the SMiRL code (Later, this can be any python code you want to run on the cluster). 
  b) Clone the doodad codebase as well.
  c) Setup your [doodad config](https://github.com/Neo-X/SMiRL_Code/blob/master/launchers/config.py). You will need to update some of the paths for your local computer
  d) You should follow [the instructions](https://github.com/Neo-X/SMiRL_Code#build-instruction) for setting up SMiRL. Create a conda environment for SMiRL and install SMiRL and doodad, etc.
  e) Run a [basic SMiRL training](https://github.com/Neo-X/SMiRL_Code#commands) simulation. Now the code should be working on your computer, and you have some familiarity with configuring doodad to run locally.
2. Running the SMiRL code in docker on your local computer.
  a) To build the docker image you can run [this script](https://github.com/Neo-X/SMiRL_Code/blob/master/build_push_docker.sh) in the smirl code that creates the docker container from the [Dockerfile](https://github.com/Neo-X/SMiRL_Code/blob/master/Dockerfile). You may need to create an account on dockerhub.com to push your container so others (including yourself) can access the docker container.
  b) I have created and pushed this docker file already. You can find it on my [dockerhub.com](https://hub.docker.com/repository/docker/gberseth/smirl).
  c) Update your `launchers/config_private.py` to use the docker container you created. For me this is 
  `DOODAD_DOCKER_IMAGE = 'gberseth/smirl:latest'`
  d) Now you should be able to run the SMiRL code inside the docker contain with the command

  ```
python3 scripts/dqn_smirl.py --config=configs/tetris_SMiRL.json --run_mode=local_docker --exp_name=test_smirl
```
3. Running your code in singularity. This part is easy to perform on the lab computers where apptainer and singularity are already installed (singularity is a bit annoying to install).
  a) Build the singularity image from the docker container using apptainer and [this script](https://github.com/Neo-X/SMiRL_Code/blob/master/build_apptainer.sh). You may need to [install apptainer](https://github.com/apptainer/apptainer/releases/tag/v1.0.2). 
  b) Update the `launchers/config_private.py` file so the code can find the singularity image for the new apptainer image.
` SINGULARITY_IMAGE = 'smirl_app.sif' `
  c)  Now you can run the code in singularity.
  ```
python3 scripts/dqn_smirl.py --config=configs/tetris_SMiRL.json --run_mode=local_singularity --exp_name=test_smirl
```
4. Now to move to Compute Canada (Or anywhere you want to run your code)
  a) Make sure you can login to compute Canada. I am using  `ssh gberseth@cedar.computecanada.ca`. YOu can add you RSA key to your compute Canada profile for easy login.
  b) Copy your singularity image to compute Canada (Globus might be faster, but this worked well enough). 
   ```
   scp smirl_app.sif  $USER@cedar.computecanada.ca:
   ```
  c) Clone your code to compute Canada.

- I cloned doodad and SMiRL to `/home/gberseth/playground`
- Also clone SMiRL_Code to `~/projects/def-gberseth/gberseth/playground/`. This is because we are not allowed to run Slurm jobs from our user folders on cedar.

d) make a virtual environment on compute Canada (in `~/projects/playground/`). Compute canada recomand virtual environments over conda.

 - load the python module on the server, `module load python/3.7`.
 - Using command `virtualenv --no-download smirl`
 - activate the virtual environment `source smirl2/bin/activate`
 
  e) Install doodad and SMiRL_Code with `pip install -e ./` in the folders where the code is stored.
  f) You can not run singularity locally on the login node (WHICH YOU SHOULD NEVER DO IN PRACTICE), but we will do just to check that the code runs.
 - Create another `launchers/config_private.py` file on compute Canada. You can copy and paste your local version on your computer; however, you will have to make a few updates.
 - After the launcher is configured properly, you can run the code from Smirl in ~/projects/def-gberseth/$USER/playground/ ( Note: You may need to change your cloudpickle version using `pip install cloudpickle==1.3.0`.)

  ```python3 scripts/dqn_smirl.py --config=configs/tetris_SMiRL.json --run_mode=local_singularity --exp_name=test_smirl```

 - Let this run for 2-3 iterations (1 minute) to make sure things work well.

5. Put your experiment on the Slurm job queue. This is how you will normally train ML models.
  a) Using the command
```
python3 scripts/dqn_smirl.py --config=configs/tetris_SMiRL.json --run_mode=slurm_singularity --exp_name=test_smirl
```
  b) You can see the current status of your jobs using `squeue -u $USER`
  c) The code will write the job output to `cat slurm-<job_id>.out`

This is the final launcher config I am using on CC.
[config_private_cc.py.txt](https://github.com/montrealrobotics/doodad/files/9972502/config_private_cc.py.txt)

Let me know if you have questions.

**Comment by Neo-X** [comment]
For installing 
```
#!/bin/bash
###
### Print list of resouce links so you can quickly look through them
###

sudo apt-get -y install curl

curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo apt-key add -

sudo add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/ubuntu $(lsb_release -cs) stable"

sudo apt-get update

apt-cache policy docker-ce

sudo apt-get install -y docker-ce

sudo systemctl status docker

distribution=$(. /etc/os-release;echo $ID$VERSION_ID) \
   && curl -s -L https://nvidia.github.io/nvidia-docker/gpgkey | sudo apt-key add - \
   && curl -s -L https://nvidia.github.io/nvidia-docker/$distribution/nvidia-docker.list | sudo tee /etc/apt/sources.list.d/nvidia-docker.list

sudo apt-get update

sudo apt-get install -y nvidia-docker2

sudo usermod -aG docker ${USER}

su - ${USER}

id -nG

sudo systemctl restart docker

sudo docker run --rm --gpus all nvidia/cuda:10.1-base nvidia-smi

### Setup GCP authorization

# Add the Cloud SDK distribution URI as a package source
echo "deb [signed-by=/usr/share/keyrings/cloud.google.gpg] http://packages.cloud.google.com/apt cloud-sdk main" | sudo tee -a /etc/apt/sources.list.d/google-cloud-sdk.list

# Import the Google Cloud Platform public key
curl https://packages.cloud.google.com/apt/doc/apt-key.gpg | sudo apt-key --keyring /usr/share/keyrings/cloud.google.gpg add -

# Update the package list and install the Cloud SDK
sudo apt-get update && sudo apt-get install google-cloud-sdk

### init and login
gcloud init

### Setup docker cridentials
gcloud auth configure-docker
```

**Comment by Neo-X** [comment]
To use doodad for your own code you can look at this wrapper function.
https://github.com/Neo-X/SMiRL_Code/blob/master/util/tuneParams.py#L29

**Comment by kirstyellis** [comment]
part 2c "Update your launchers/config_private.py", also add the path to your code in CODE_DIRS_TO_MOUNT. I didn't and got hit with a
ModuleNotFoundError: No module named 'surprise'
